packages/amplify-graphql-index-transformer/src/utils.ts

```diff
@@ -50,3 +51,60 @@ export function validateNotSelfReferencing(config: IndexDirectiveConfiguration |
     }
   }
 }
+
+/**
+ * Checks if @auth owner field has been set to a sortKeyField for @primaryKey
+ */
+export const validateNotOwnerAuth = (
+  sortKeyField: string,
+  { object }: PrimaryKeyDirectiveConfiguration,
+  ctx: TransformerContextProvider,
+): boolean => {
+  const authDir = object.directives?.find(dir => dir.name.value === 'auth');
+
+  if (!authDir) return true;
+
+  const dirRules = (authDir.arguments?.find(arg => arg.name.value === 'rules')?.value as ListValueNode)?.values || [];
+  const hasOwnerFieldAsSortKey = dirRules.some(rule => {
+    let isOwner = false;
+    let identityClaimIsSet = false;
+    let ownerFieldIsSet = false;
+    let usesMultiClaim = false;
+    let sortKeyFieldIsAuthField = false;
```

Comment:
I'll OOS for this PR, since this isn't yet an established pattern, but I prefer to avoid let wherever possible, and rely instead on returning consistent shapes, and reducing state where necessary. In general for something like this I'd probably refactor to have your inner loop (i.e. `rule.fields?.forEach` turn into a `rule.fields.map(doesRuleHaveOwnerForSortKeyField)`, then just return a `.some` for that mapped value. As written, I'm not sure if it's possible to have `sortKeyFieldIsAuthField` be set to `true` for a single rule, and `usesImplicitOwnerField` for another unrelated rule, and result in a false positive here.

Reply:
As written, there's some chance for state interaction between rules, which I don't think is by design. If we need rules to 'see' each other, we should make that super explicit, and either preprocess the necessary cross-rule state, or otherwise allow the rule checker to look at it's neighbors rather than relying on global state.

Reply:
I'm fine with that. This was the best way I could configure for my own readability since the GQL objects can be messy to parse (I always wanted to work on some GQL adapters to integrate the code into the transformers), but I can change the pattern. I'll refactor this.